### DISCUSSION TITLE
Access SyntaxTree from another project in solution inside a SourceGenerator

### DISCUSSION THREAD

**Discussion by rafalkwol** [dotnet/roslyn#50874] [Q&A] [answered]
Hello,

I have a problem I can't seem to solve at the moment with source generators, but maybe I'm just missing something. Let's say I have project named "SourceGenerator" which has the source generator class. Than I also have a project called "Model" which references the "SourceGenerator" project. There is also third project called "Repository". What I want to achieve is to use the source generator class to generate automatic repository classes basing it on the models in the "Model" project, but not include them in the "Model" project, but in the "Repository" project. Is it possible to achieve that?

Another way that would be acceptable would be to add reference to "SourceGenerator" project in the "Repository" project instead of "Model" project, but then how do I access the the SyntaxTree of "Model" project in the source generator class.

The only solution I found is to read all the source files from another project from disk, change them into a compilation object and then analyze syntax trees. It works, but it doesn't use the benefit of syntax visitor as for every change in the code it loads all the files, analyzes them and generates source code. It doesn't seem right.

Any ideas how to solve this issue? Thanks in advance. :)

**Comment by jasonmalinowski** [comment, marked as answer]
> Another way that would be acceptable would be to add reference to "SourceGenerator" project in the "Repository" project instead of "Model" project, but then how do I access the the SyntaxTree of "Model" project in the source generator class.

So this is probably the right approach, but yes the source generator running against the Repository project can't see the _syntax trees_ of the Model project because it's already been built and compiled. What it _can_ see though is symbols: if the Repository project has a reference to the Model project the types in the Model project will be visible through the symbol APIs. You could then look at classes/methods that way, as long as you don't need to look at the code inside those methods.

**Reply by rafalkwol** [reply]
That seems like the way to go. Any tips how I can access referenced assembly symbols from the GeneratorExecutionContext object? As I can't seem to find an easy way to do it - or do I have to do it with using some external namespace? I'm pretty new to Roslyn and it's codebase so I'm not really sure what the right direction is here.

**Reply by jasonmalinowski** [reply]
So the GeneratorExecutionContext will give you a [compilation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.compilation?view=roslyn-dotnet), from there you can grab Compilation.GlobalNamespace which will let you see the top level namespace. From there you can start recursing through the child namespaces and then get to types. If you are able to filter right away to the correct metadata reference, you can also call [this API](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.compilation.getassemblyormodulesymbol?view=roslyn-dotnet#Microsoft_CodeAnalysis_Compilation_GetAssemblyOrModuleSymbol_Microsoft_CodeAnalysis_MetadataReference_) which would give you an IAssemblySymbol which then has a GlobalNamespace property that will give you types filtered to just that global namespace.

**Reply by rafalkwol** [reply]
Nice, that's exactly what I needed to start. Thanks a lot for help.

**Reply by onetsmr** [reply]
Thanks, it helped me. I have main web project with controllers (and link to generator project here) and the secondary project with models. Compilation.GlobalNamespace also has namespaces and types from compilling project. So I can get symbols for controller classes from Web project.
